Refactor Mirabuf Loading System `[AARD-2060]`

rutmanz opened the pull request:
## Task

<!--
Please include any relevant Jira ticket ID(s) at the end of the PR title, in the form AARD-xxxx, where "AARD" is Jira project.
Include the same Jira ticket ID(s) in this section.
-->

AARD-2060 - Refactor mirabuf caching service


AARD-2069 - Make uploading local mirabuf files enforce correct miratype (with gamepieces we can't just infer miratype completely)
<!--
Provide a brief description of what the task was here.
-->

## Symptom
<!--
How does the problem manifest itself?
Describe the problem as seen by the user, or by the caller or the code if it's not directly visible to the user.

Note: "Symptom" can include new product use cases, not just bugs.
-->

The current MirabufCachingService made a few decisions that made it hard to understand and had some key limitations:
- The `id` field was unique but not consistent (it was based on the time that you first downloaded the file to your cache). This was the impetus for changing this, as we needed consistent IDs for multiplayer.
- The `cacheKey` field was either the name of the APS project, the remote download URL, or the hash of the locally uploaded file
- The `MirabufCacheInfo` type was used both to store information about the OPFS files and to store in-memory references to the ArrayBuffers
- Everything was split by type, so there were two versions of every datastructure, one for robots, one for fields (and gamepieces added a third). There's no actual reason to require this as one model should never appear as both a field and a robot
- There was a bunch of duplicate code
- It didn't actually work when OPFS was unavailable

## Solution

<!--
How did you fix the problem/symptom?
Explain your approach and reasoning for choosing this solution.
-->

- When OPFS is unavailable (some old browsers, insecure contexts), it will gracefully fall back and avoid producing errors.
- The `id` field has been replaced with a `hash` field, that is the SHA-1 hash of the assembly (consistent and unique)
- Removed unnecesssary and duplicate code
- Uses one map and OPFS folder for all assets
- Checks on load that all the assets are still there and removes any extras
- Fields aren't randomly optional (looking at you `name`), and there is a single source of truth for asset metadata
## Verification

> [!NOTE]  
> Your mirabuf cache will be cleared every time you switch between Synthesis on this branch and Synthesis not on this branch

- Test that downloading mirabuf assets works (and stays after reload)
- Test that uploading mirabuf assets works (and stays after reload)
- Test that downloading APS mirabuf assets works (and stays after reload)
- Test that deleting assets works and [removes them from OPFS](https://chromewebstore.google.com/detail/opfs-viewer/bebjgdnmkhibekhoijhhbdpfdddpefci)
- Test that messing around in your browser's dev tools and deleting files/localstorage items doesn't cause undefined behavior
<!--
How did you test and verify your changes were correct?
List steps taken, tests/added/updated, and any manual verification done that should be replicated in review.
-->

---

Before merging, ensure the following criteria are met:

- [ ] All acceptance criteria outlined in the ticket are met.
- [ ] Necessary test cases have been added and updated.
- [ ] A feature toggle or safe disable path has been added (if applicable).
- [ ] User-facing polish:
  - Ask: *"Is this ready-looking?"*
- [ ] Cross-linking between Jira and GitHub:
  - PR links to the relevant Jira issue.
  - Jira ticket has a comment referencing this PR.

rutmanz committed:
feat: refactor mirabuf caching service

rutmanz committed:
fix: unit tests

rutmanz committed:
fix: make old caches not die

rutmanz committed:
fix: also remove old localstorage keys

rutmanz committed:
feat: allow loading directly from memory cache as fallback

rutmanz committed:
fix: make memory cache actually used

rutmanz committed:
Merge remote-tracking branch 'origin/dev' into zachr/2060/mirabuf-refactor

rutmanz committed:
fix: unit tests

rutmanz committed:
fix: merge conflicts

rutmanz committed:
feat: prevent uploading local files as wrong mira type

rutmanz committed:
feat: remove files that don't exist in OPFS on load

rutmanz committed:
feat: allow mirabuf caching service to work in insecure contexts

azaleacolburn commented on fission/src/ui/panels/DeveloperToolPanel.tsx:
```diff
@@ -10,9 +9,32 @@ import World from "@/systems/World"
 
+async function saveToCache() {
+    const field = World.sceneRenderer.mirabufSceneObjects.getField()
+    if (field) {
```
this would be a good place to early-return

rutmanz committed:
refactor: early return

rutmanz committed:
feat: make manifest system more robust

rutmanz committed:
Merge remote-tracking branch 'origin/dev' into zachr/2060/mirabuf-refactor

rutmanz committed:
fix: ts configs

fix: try to make unit tests work

fix: actually make unit tests work

feat: add unit tests for default assets

fix: tsconfig issues again

PepperLola reviewed (changes requested):
Do the import reordering changes need to be included in this PR? ~Also, it doesn't load the default robots and~ clicking the download all button spawns a progress popup that has a broken format

<img width="244" height="90" alt="Image" src="https://github.com/user-attachments/assets/33eeb5a3-2994-4dd6-8c29-a730c6908c00" />

EDIT: loading default robots was because I hadn't run the `assetpack` command

PepperLola commented on fission/src/ui/panels/mirabuf/ImportMirabufPanel.tsx:
```diff
@@ -295,114 +220,128 @@ const ImportMirabufPanel: React.FC<PanelImplProps<void, ImportMirabufPanelCustom
+
+            let completeCount = 0
+            const totalCount = toCache.length
+            status.update(`Downloading... (0/${totalCount})}`, 0.05)
```
```suggestion
            status.update(`Downloading... (0/${totalCount})`, 0.05)
```

rutmanz committed:
fix: undo import sorting

rutmanz committed:
Update fission/src/ui/panels/mirabuf/ImportMirabufPanel.tsx

Co-authored-by: Julian Wright <PepperLola@users.noreply.github.com>

PepperLola commented on fission/src/ui/panels/mirabuf/ImportMirabufPanel.tsx:
```diff
@@ -295,114 +220,128 @@ const ImportMirabufPanel: React.FC<PanelImplProps<void, ImportMirabufPanelCustom
+                                status.done()
+                            } else {
+                                status.update(
+                                    `Downloading... (${completeCount}/${totalCount})}`,
```
there was another
```suggestion
                                    `Downloading... (${completeCount}/${totalCount})`,
```

PepperLola reviewed (changes requested):
one more thing 🍎

rutmanz committed:
Update fission/src/ui/panels/mirabuf/ImportMirabufPanel.tsx

Co-authored-by: Julian Wright <PepperLola@users.noreply.github.com>

rutmanz committed:
fix: prevent unit tests timing out

rutmanz commented:
- [x] Make assetpack tests run separately and not as part of unit test workflow

rutmanz committed:
Merge remote-tracking branch 'origin/dev' into zachr/2060/mirabuf-refactor

rutmanz committed:
fix: don't run hash tests, add test for removal

rutmanz committed:
fix: remove debug tool

rutmanz committed:
Merge remote-tracking branch 'origin/dev' into zachr/2060/mirabuf-refactor

rutmanz committed:
fix: make unit tests pass

PepperLola reviewed (approved):
Looks good

rutmanz committed:
feat: add tests for assetpack as separate workflow

rutmanz committed:
Downgrade playwight version because github actions doesn't know about the new one

BrandonPacewic reviewed (approved):
I understand we are try to fix the workflow. Everything other than that looks good to me. Waiting for final changes.

rutmanz committed:
fix: make env variable be exposed
